core/src/test/java/org/apache/iceberg/rest/TestHTTPClient.java

```diff
@@ -121,6 +128,108 @@ public void testHeadFailure() throws JsonProcessingException {
     testHttpMethodOnFailure(HttpMethod.HEAD);
   }
 
+  @Test
+  public void testProxyServer() throws IOException {
+    int proxyPort = 1070;
+    try (ClientAndServer proxyServer = startClientAndServer(proxyPort);
+        RESTClient clientWithProxy =
+            HTTPClient.builder(ImmutableMap.of())
+                .uri(URI)
+                .withProxy("localhost", proxyPort)
+                .build()) {
+      String path = "v1/config";
+
+      HttpRequest mockRequest =
+          request("/" + path).withMethod(HttpMethod.HEAD.name().toUpperCase(Locale.ROOT));
+
+      HttpResponse mockResponse = response().withStatusCode(200);
+
+      mockServer.when(mockRequest).respond(mockResponse);
+      proxyServer.when(mockRequest).respond(mockResponse);
+
+      clientWithProxy.head(path, ImmutableMap.of(), (onError) -> {});
+      proxyServer.verify(mockRequest, VerificationTimes.exactly(1));
+    }
+  }
+
+  @Test
+  public void testProxyCredentialProviderWithoutProxyServer() {
+    HttpHost proxy = new HttpHost("localhost", 1070);
+    BasicCredentialsProvider credentialsProvider = new BasicCredentialsProvider();
+    credentialsProvider.setCredentials(
+        new AuthScope(proxy),
+        new UsernamePasswordCredentials("test-username", "test-password".toCharArray()));
+    Assertions.assertThatThrownBy(
+            () -> {
+              HTTPClient.builder(ImmutableMap.of())
+                  .uri(URI)
+                  .withProxyCredentialsProvider(credentialsProvider)
+                  .build();
+            })
+        .isInstanceOf(NullPointerException.class)
+        .hasMessage("Invalid http client proxy for proxy credentials provider: null");
+  }
+
+  @Test
+  public void testProxyServerWithNullHostname() {
+    Assertions.assertThatThrownBy(
+            () -> {
+              HTTPClient.builder(ImmutableMap.of()).uri(URI).withProxy(null, 1070).build();
+            })
+        .isInstanceOf(NullPointerException.class)
+        .hasMessage("Invalid hostname for http client proxy: null");
+  }
+
+  @Test
+  public void testProxyAuthenticationFailure() throws IOException {
+    int proxyPort = 1070;
```

Comment:
you might want to pick a different port for tests that actually start a server.  If tests ever run in parallel, then any of the tests that use the same port could randomly fail